### PR TITLE
bsi v2

### PR DESCRIPTION
**Pull request by riteshnoronha**
Add compliance doc for bsi - v2.0 for review.

### CODE REVIEW

**Comment by riteshnoronha** [comment]
@fvsamson  @surendrapathak  I have read & re-read the BSI v2.0 spec multiple times and i have come up with the compliance doc, before i start the implementations.  Wanted to call out a couple of things 

- CDX minimum version requirement from 1.4 to 1.5 was called out, but SPDX 2.3 -> 2.2.1 was not. 
- Added column for SPDX 3.0, will fill that out later once i complete the golang library for SPDX 3.0 parsing. 
- Need help with Executable/Archive and structured properties, not sure how these translate. 
- BSI has reserved namespace for CDX properties, can they be leveraged for custom data? 
- Associated Licenses: Seems like a new concept, but my interpretation is more on the lines of License Expression should be present, not list of licenses for concluded licenses. 
- Feels like "Additional SBOM & Component Fields" are almost mandatory now, in v1.1 it was optional ? 
- SBOM's should not contain vulns (Love it)

**Comment by viveksahu26** [comment]
Yeah, the compliance doc looks good. https://github.com/interlynk-io/sbomqs/blob/feature/bsi-v2/Compliance.md . 
Some of the fields as mentioned above like Archive/executable and structured looks in the sboms.

**Comment by fvsamson** [comment]
Wow, you are quick.  I am currently travelling, so my time is a bit limited and hence at BSI our weekly TR-03183-2 writers team teleconference was cancelled.  What I can provide the next days are some replies to @riteshnoronha's points above, hopefully a first proper review of your compliance document sometime this week (just glanced over it now), and I will discuss both (questions and document) with my colleagues at our teleconference next week.

Side note: We plan to publish our thinking of mapping TR-03183-2's requirements to CDX and SPDX tags in v2.1.0 of TR-03183-2.  We are aware, that for a few requirements there are no appropriate tags, but there are a couple of generic ones one can use as `name:value` fields.  Likely we will also require at least SPDX 3.0 (and presumably CDX 1.6), because some new tags make this mapping a lot easier.

This mapping task is primarily performed by a colleague who also has become the one most savvy of us in the CDX and SPDX specifications.  I will ask him to chime in here next week.

File: Compliance.md [hunk 1]
```diff
@@ -1,86 +1,134 @@
...
+|                                               | `dependencies`               | dependencies, compositions                                             | relationships                                                              | TBD       | cdx we look for attestations via compositions, spdx nothing exists                                                                             |
+|                                               | `associated license`         | component->license->Expression                                         | packageConcluded                                                           | TBD       | we lookup sdpx,spdx-exceptions,aboutcode, and licenseRef-                                                                                      |
+|                                               | `hash`                       | component->hashes                                                      | package->checksums                                                         | TBD       | we only look for sha-256                                                                                                                       |
```
Comment (lpanni) [inline Compliance.md:32]: the guideline specified SHA-512 for "hash value of the deployable component"

**Review by lpanni — COMMENTED** [review]
There seems to be a slight difference in the meaning between `associated license` and `concluded license`.  But in a multi-licensing example as listed in Annex 8.1.9, I'd consider the choice for the associated license also to be the concluded license.
Who makes the decision seems to be the difference: the SBOM creator (associated) or the component licensee (concluded), which will be the SBOM creator for any direct dependencies.

Example:
- Direct dependency to a component A that declares MIT OR GPL-2.0-only => SBOM creator can decide to use MIT and list that as concluded
- Transitive dependency to the same component A via a directy dependency component B declaring only MIT => creator of component B already made the decision to use MIT, so the SBOM creator has to list MIT as associated license instead of concluded.

Am I right on this one or does it even make a difference in practice?

**Comment by surendrapathak** [comment]
**Notes:**

- 3.1 Definition of SBOM - specifically points out `An SBOM MUST NOT contain vulnerability information.` Therefore, I recommend removing that row.
4 SBOM Formats—limits both specifications and specification versions BUT ALSO file formats to JSON/XML (removing RDF and TV). sbomqs should add a row for that.
- Typo in 5.2.2 `originatior`
- On `associated license`, I agree with @lpanni 's point above and also from the description itself : ```Associated licence(s) of the component from the perspective of the
SBOM creator. For specifics, see sections 6.1 and 8.1.9.```. Specifically, the phrase indicates a choice made by the SBOM creator (`concluded`) rather than just the listed (`associated`). However, if the choice has not been made (concluded license not present), then it seems safe to fall back to associated as per section 8.1.9 `Associated licences are all licences under which a component can be used by the licensee.`
- 5.2.2 `hashes` - Since the requirements limits to `deployed/deployable component`, we should limit this check to the specific component types (not applicable to say ARCHIVE , TEXT, VIDEO etc)

Rest looks good to me.

**Comment by riteshnoronha** [comment]
@fvsamson i am going to start development based on my initial take, we can refine it once you have bandwidth to review it.

**Review by viveksahu26 — APPROVED** [review]

**Comment by fvsamson** [comment]
I am terribly sorry to have stalled this for so long.  November last year was packed, because I had to use up some of my off-days before the end of the year, hence I worked only very few days in December.   But since mid-January I am thinking of this almost weekly, considering "the next time I have at least a fully day to rehearse these questions, and to research and provide answers, just do it"; well that "at least a day time I am not occupied with appointments etc." never came.
This week was the first significantly quieter week for a long time, and I decided to address the questions one-by-one, but not necessarily in one round (even though I still believe this would be the most efficient approach: though an unsuitable one if a precondition is never fulfilled).
Another contributing factor is that we are close to release v2.1.0, for which I had to review and reconsider most of the topics you addressed, anyway.  Hence I should be able to provide concise answers quicker and easier (i.e. research less what we wrote and less time to consider an determine the best passages to quote for you).

WRT v2.1.0: This turned out to be a continuation of v2.0.0, as planned, i.e. we have altered almost nothing, rephrased a couple of statements to increase legibility, and the only new thing is a mapping of our requirements to SPDX- / CycloneDX-tags, which should ease your work a lot.

**Comment by fvsamson** [comment]
@surendrapathak wrote [2024-10-18](https://github.com/interlynk-io/sbomqs/pull/331#issuecomment-2421727158):

> **Notes:**
> 
> * 3.1 Definition of SBOM - specifically points out `An SBOM MUST NOT contain vulnerability information.`

... or we do not call it "SBOM" any longer, but "file which contains SBOM and vulnerability data".  Please note that such a file can still be conformant to BSI TR-03183-2.

> Therefore, I recommend removing that row.

I have not looked at it yet, but this consequence may be too harsh.

> 4 SBOM Formats—limits both specifications and specification versions BUT ALSO file formats to JSON/XML (removing RDF and TV).

That is correct and explicitly done deliberately.

> * On `associated license`, I agree with @lpanni 's point above and also from the description itself : `Associated licence(s) of the component from the perspective of the SBOM creator. For specifics, see sections 6.1 and 8.1.9.`. Specifically, the phrase indicates a choice made by the SBOM creator (`concluded`) rather than just the listed (`associated`). However, if the choice has not been made (concluded license not present), then it seems safe to fall back to associated as per section 8.1.9 `Associated licences are all licences under which a component can be used by the licensee.`

Basically correct, except for modelling that the SBOM creator has any choice but to record what is there (see [here for details](https://github.com/interlynk-io/sbomqs/pull/331#issuecomment-2946537806)).

> * 5.2.2 `hashes` - Since the requirements limits to `deployed/deployable component`, we should limit this check to the specific component types (not applicable to say ARCHIVE , TEXT, VIDEO etc)

But aren't these artefacts as binaries?  What is the fundamental difference between a text-based default configuration file and man(ual) page file?  All these become deployed upon installation on local mass storage.  But maybe I am failing to see a relevant aspect; if so, please denote and rectify my misconception.

**Comment by fvsamson** [comment]
Oh, GitHub's "Review" functionality, which seems to be broken for a while.  I love(ed) it (and apparently @lpanni, too), because it allows for discussion threads which are attached to (a) specific code line(s).  But using my private GitHub account, I discovered months ago, that these comments are not visible for some.  Here I can see my three review-comments (created between 2024-10-12 and 2024-10-14) as well as one by @lpanni to which I replied today; but it seems that none of the others noticed them, @riteshnoronha, @viveksahu26 and @surendrapathak?

---------------------------------

## Copies of extant review comments as of 2025-06-05

---------------------------------

[**`Compliance.md`**, line 4](https://github.com/interlynk-io/sbomqs/pull/331/files/206f4d580fa942a762487756bdcebcb50fe2f7a5#diff-ed1d122f0001e698a225da370f48fb661842e352e1aa8bdfc0720d4a203e099bR4)

@fvsamson: I initially missed this one: Please append `-2` and keep the `v2.0` / `v2.0.0` consistent with line 12.

---------------------------------

[**`Compliance.md`**, line 8](https://github.com/interlynk-io/sbomqs/pull/331/files/206f4d580fa942a762487756bdcebcb50fe2f7a5#diff-ed1d122f0001e698a225da370f48fb661842e352e1aa8bdfc0720d4a203e099bR8)

@fvsamson:
Suggestion:
`## BSI TR-03183-2: SBOM Technical Guideline by BSI`

Reasoning:
- Please always explicitly refer to part 2 (hence the `-2`), because public drafts of parts 1 and 3 are now also published, which do focus on other aspects of implementing requirements EU's CRA (Cyber Resilience Act) than SBOMs.
- We call each such document a Technical Guideline (i.e. in singular).
- [not strictly necessary] The preferred notation is "BSI TR-xxxxx", though I tend to omit the "BSI" after the first occurrence of that term (to be optionally repeated again somewhere after a couple of occurrences).  Well, this is its first occurrence in this document (or at least in this section, as it is a section header), hence I put it in.
- You may try to find a location where you can (once) expand "BSI" to "German Federal Office for Information Security (BSI)" (or the other way around, if that fits better: "BSI (the German Federal Office for Information Security") in order to avoid the easy misinterpretation as the "British Standards Institute (BSI)", which also provides guidelines and specifications for information technology (but also many other technical fields, in contrast to us).

---------------------------------

[**`Compliance.md`**, line 12](https://github.com/interlynk-io/sbomqs/pull/331/files/206f4d580fa942a762487756bdcebcb50fe2f7a5#diff-ed1d122f0001e698a225da370f48fb661842e352e1aa8bdfc0720d4a203e099bR12)

@fvsamson:
We switched to Semantic Versioning (see [SemVer.org](https://semver.org/)) by version 2.0.0, i.e. three numerical fields with clearly defined meaning (i.e., when which field is increased).  As that implies that the third (i.e. last) field is only altered for bug fixes, which do not alter meaning, APIs etc. (e.g. spelling and grammar fixes), it might make sense to refer to "v2.0" (as you already do) or "v2.0.x".

**But** the web-link you use explicitly points to v2.0.0.  Hence I suggest to use the "short link" https://www.bsi.bund.de/dok/TR-03183-en (which redirects to horrible https://www.bsi.bund.de/EN/Themen/Unternehmen-und-Organisationen/Standards-und-Zertifizierung/Technische-Richtlinien/TR-nach-Thema-sortiert/tr03183/TR-03183_node.html) which lists the last two versions of all three parts of TR-03183.

While I prefer this generic solution (needs less maintenance etc.), alternatively you may keep the link as is and call it "v2.0.0" here and also in line 4.

---------------------------------

[**`Compliance.md`**, line 32](https://github.com/interlynk-io/sbomqs/pull/331/files/206f4d580fa942a762487756bdcebcb50fe2f7a5#diff-ed1d122f0001e698a225da370f48fb661842e352e1aa8bdfc0720d4a203e099bR32)

@lpanni: the guideline specified SHA-512 for "hash value of the deployable component"

@fvsamson:
Yes, I switched to SHA-512 for v2.0.0, because it is ~ 1,5 times faster on 64-bit CPUs for files lager than 1024 bits (128 Bytes) than SHA-256.

Side note: I have been asked so many times, why we do not use [OmniBOR](https://github.com/omnibor/spec/tree/main/spec) for artefact hashing, especially for hashing source code.  I really would like to, but it is far from being ready (in the sense of being a practically usable, concise and consistent specification).
Background: Omnibor tries to recycle SHA-256 Git Object IDs (Git OIDs, which actually hash a little more than just a file's content), which currently comprise either a SHA-1 or SHA-256 hash values set ("in stone") when initialising a Git repository.  The Git developers are working on a migration scheme ... for a decade (and counting).  All major source code hosting platforms based on Git (GitHub, Gitlab and Forgejo used by e.g. Codeberg.org) still initialise new repositories with SHA-1 OIDs, because AFAIK one still even cannot pose PRs between Git repos using different hash algorithms.  Hence, besides the OmniBOR specification being far from ready to be applied practically, the elegance of reusing extant (i.e. already calculated) hash values is a promise which cannot be fulfilled, yet and in the next couple of years).  Note that multiple hash values can be provided for each artefact in an SBOM, thus using multiple schemes concurrently is technically feasible, albeit computationally expensive.

**Comment by fvsamson** [comment]
> ~…, so you can evaluate, if you still want to finish this based on v2.0.0 or if you prefer to address v2.1.0 right away; I assume the second approach results in slightly less work (in the sense of "required efforts")~ _(Edit2: That was a "bad idea"™, …

… as I understood now, that I rehearsed what we specify in the section "Transitional scheme": A software product with an SBOM which was once conformant to a then recent version of BSI TR-03183-2 (with a 6 month transitional period of what "recent version" means, after a new version was released) will stay conformant indefinitely, as long as that software product is not updated (and hence its SBOM, too).
Thus it is almost equally much of interest to be able to check conformance to an older version of BSI TR-03183-2 than to the recent version of it.

Consequently I kindly ask you to finish your checks for v2.0.0 (and will try to provide a next round of QS, then), before you implement the checks for v2.1.0 on top of that.  As denoted earlier, you should be able to utilise the (hopefully) improved normative texts and explanations in v2.1.0-beta3 to ease your understanding of v2.0.0 (but please check if a specific, altered text in v2.1.0 is not related to one of the few technical things we did change from v2.0.0 to v2.1.0): If still linguistic or technical ambiguities remain from your perspective, please ask specific questions here or via BSI's contact address for all three parts of TR-03183: \<tr03183@bsi.bund.de\>
<sub>BTW, if I ever do not respond for multiple months again, please do address [my three colleagues in our development team for the TR-03183-s](https://github.com/orgs/tr-03183/people?query=role%3Aowner) (Thomas Seider, Michael Schuster and Thomas Schmidt) by "**`@`**"-mentions; further steps might be to open [an issue at our development repository for BSI's CDX property taxonomy](https://github.com/tr-03183/tr-03183-cyclonedx-property-taxonomy/issues) or to email to aforementioned group account for the TR-03183-s.</sub>

-------------------------------------------------------------

@riteshnoronha wrote [on 2025-06-07](https://github.com/interlynk-io/sbomqs/pull/331#issuecomment-2951327754): 
> We are also working on sbomqs summary commands for bsi-v1 & bsi-v2. This can hopefully give users a quick overview of their sboms adherence to the Technical guides.
> 
> `sbomqs score -c bsi-v1 sbom.json`
> `sbomqs score -c bsi-v2 sbom.json`
> `sbomqs score -c bsi-v2.1.0 sbom.json`

The exemplary output of your `sbomqs score` sub-command looks great (I have a few ideas for minor improvements, but will try to evaluate in the next weeks if I rather directly pose a PR or write up suggestions for improvement here).

What I am not really happy about are some of the parameters to the "**`-c`**"-option: Please replace `bsi-v1` by `bsi-v1.1` and `bsi-v2` by `bsi-v2.0.0`.  If you already released a version with these original option parameters (hence someone may already utilise it in an automated (e.g. scripted) manner, please keep the old parameters `bsi-v1` and `bsi-v2` as undocumented alias names for their new counterparts.
Thanks!

**Comment by riteshnoronha** [comment]
@fvsamson yes the "Transitional scheme" is why we are keeping the implementations around so users are not left with only the latest version, due to the reasons stated  above. 

The sbomqs summary commands is in development now, so we welcome your feedback. We can make the changes to the parameters. 

```
sbomqs score -c bsi-v1.1.0 sbom.json
sbomqs score -c bsi-v2.0.0 sbom.json
sbomqs score -c bsi-v2.1.0 sbom.json
``

**Comment by fvsamson** [comment]
Dear Ritesh, we were not as consistent as you seem to assume. :wink:  We really started off with a two field versioning scheme for v1.0 and v1.1 and then switched to Semantic Versioning, i.e. three numeric fields.  Hence IMO it ought to be:
```
sbomqs score -c bsi-v1.1 sbom.json
sbomqs score -c bsi-v2.0.0 sbom.json
sbomqs score -c bsi-v2.1.0 sbom.json
```
<sup>**Background:** I asked after the release of TR-03813-2 v1.0 in one of our internal development meetings, why we strongly suggest Semantic Versioning *in* this Technical Guideline, but do not use SemVer *for* this Technical Guideline.  Actually this resulted in more discussion than I expected ("all other BSI TRs do not use SemVer" (at that time), "this is a TR not an SBOM" etc.) but lastly my arguments were convincing a majority.  But as we already started v1.1 and its development scope was primarily to translate v1.0 to English (lastly it turned out to be slightly more than that, due to BSI-external feedback on content), we decided to perform the switch to SemVer with v2.0.0.</sup>

**Comment by riteshnoronha** [comment]
@fvsamson we have gone ahead with the simplified scheme. We are almost ready to release this changes, next week.

**Comment by fvsamson** [comment]
> There seems to be a slight difference in the meaning between `associated license` and `concluded license`. But in a multi-licensing example ...

Oh, it is worse: While trying to rework the sections about licensing (in chapters 3, 5 and 8) we no believe to have comprehended that the fuzzy tag specifications of SPDX and CycloneDX using similar terms are apparently incompatible, at least in nuances.

My current approach for finalising v2.1.0 is to deliberately use completely different terms to define our requirements WRT licence categories, and then suggest a mapping to each spec (basically applying the extant scheme tom licence categories, too): Hence the tricky part is a bit deferred and its resolution may use what the specs offer (which implies "adhering to their spec-wording").  I am not happy with it, it is a bit of a bail-out of this thicket and let this issue be(come) a TR-user's issue; still it is the best (tough slightly compromised) approach I can currently think of to get v2.1.0 out of the door without contradicting or confusing wording (i.e. "terms used") compared to the CDX and SPDX specs.

**Pull request merged**